queryable-spec.ts

```diff
@@ -0,0 +1,141 @@
+
+import RDF from '@rdfjs/types';
+import {Algebra} from 'sparqlalgebrajs';
+import {EventEmitter} from 'events'; // TODO: refer to underlying interface, not to the class
+
+/*
+ * Helper union type 
+ */
+type termName = 'subject' | 'predicate' | 'object' | 'graph';
+
+/*
+ * Custom typings for the RDF/JS Stream interface as the current
+ * typings restrict the generic param Q to extensions of "BaseQuad",
+ * meaning it cannot be used for Bindings.
+ */
+export interface Stream<Q> extends EventEmitter {
+  read(): Q | null;
+}
+
+/*
+ * Map-like representation of Bindings as using plain objects could lead
+ * to collisions between variable names and object properties.
+ * Long-term goal: maintain compatibility with the native Map class.
+ */ 
+interface Bindings {
+  type: 'bindings';
+  get(variable: RDF.Variable): RDF.Term;
+  keys(): RDF.Variable[];
+  entries(): [RDF.Variable, RDF.Term][];
+  size: number;
+}
+
+/*
+ * Bindings objects are created using a dedicated factory, keeping in line
+ * with DataFactory.quad(). This also helps with facilitating support for
+ * immutability. Basic helper methods must also be provided for the most 
+ * common manipulations of bindings objects.
+ */
+interface BindingsFactory {
+  bindings(entries: [RDF.Variable, RDF.Term][]): Bindings;
+  merge(bindings: Bindings[]): Bindings;
```

Comment:
As noted by @gsvarovsky, we should document the behaviour of this method very clearly.

My thoughts on this are that all bindings must be compatible with each other (no conflicting value entries). If there are conflicts, `undefined` must be returned. (this is the behaviour in Comunica atm)